lib/hash_ring/ets.ex

```diff
@@ -0,0 +1,192 @@
+defmodule HashRing.ETS do
+  @compile :native
+
+  @default_ring_gen_gc_delay 10_000
+  @type t :: __MODULE__
+
+  use GenServer
+
+  alias HashRing.Utils
+  alias HashRing.ETS.Config
+
+  defstruct num_replicas: 0,
+            nodes: [],
+            table: nil,
+            ring_gen: 0,
+            name: nil,
+            pending_gcs: %{}
+
+  def start_link(name, nodes \\ [], num_replicas \\ 512) do
+    GenServer.start_link(__MODULE__, {name, nodes, num_replicas}, name: name)
+  end
+
+  @spec init({atom, [binary], integer}) :: t
+  def init({name, nodes, num_replicas}) do
+    table = :ets.new(name, [
+      :protected,
+      :ordered_set,
+      {:read_concurrency, true}
+    ])
+
+    state = %__MODULE__{
+      table: table,
+      num_replicas: num_replicas,
+      nodes: nodes,
+      name: name,
+    } |> rebuild
+
+    {:ok, state}
+  end
+
+  def stop(name) do
+    GenServer.stop(name)
+  end
+
+  @spec set_nodes(atom, [binary]) :: t
+  def set_nodes(name, nodes) do
+    GenServer.call(name, {:set_nodes, nodes})
+  end
+
+  @spec add_node(atom, binary) :: {:ok, t} | :error
+  def add_node(name, node) do
+    GenServer.call(name, {:add_node, node})
+  end
+
+  @spec remove_node(atom, binary) :: {:ok, t} | :error
+  def remove_node(name, node) do
+    GenServer.call(name, {:remove_node, node})
+  end
+
+  @spec get_nodes(atom) :: {:ok, [binary]} | :error
+  def get_nodes(name) do
+    GenServer.call(name, :get_nodes)
+  end
+
+  @spec force_gc(atom, integer) :: :ok | {:error, :not_pending}
+  def force_gc(name, ring_gen) do
+    GenServer.call(name, {:force_gc, ring_gen})
+  end
+
+  @spec get_ring_gen(atom) :: {:ok, integer} | :error
+  def get_ring_gen(name) do
+    with {:ok, {_, ring_gen, _}} <- Config.get(name) do
+      {:ok, ring_gen}
+    end
+  end
+
+  @spec find_node(atom, binary | integer) :: binary | nil
+  def find_node(name, key) do
```

Comment:
why do some of the public functions return `:ok` tuples and this one nil?

Reply:
this is to conform to the existing HashRing implementation.